pkg/state/invoke_applier.go

```diff
@@ -212,315 +212,333 @@ func (ia *invokeApplier) resolveAliases(actions []proto.ScriptAction, initialisa
+					return nil, false, errors.New("asset was issued by other address")
+				}
+				var n uint64
+				n, ok, err = ia.validateActionSmartAsset(a.AssetID, a, scriptPK, blockInfo, *tx.ID, tx.Timestamp, info.initialisation, acceptFailed)
+				if err != nil {
+					return nil, false, err
+				}
+				if ok {
+					scriptRuns += n
+				} else {
+					break
+				}
+				// Update asset's info
+				// Modify asset.
+				if !info.validatingUtx {
+					change := &assetBurnChange{
+						diff: int64(a.Quantity),
+					}
+					if err := ia.stor.assets.burnAsset(a.AssetID, change, info.block.ID, !info.initialisation); err != nil {
+						return nil, false, errors.Wrap(err, "failed to burn asset")
+					}
+				}
+				txDiff, err := ia.newTxDiffFromScriptBurn(scriptAddr, a)
+				if err != nil {
+					return nil, false, err
+				}
+				// diff must be saved to storage, because further asset scripts must take
+				// recent balance changes into account.
+				if err := ia.saveIntermediateDiff(txDiff); err != nil {
+					return nil, false, err
+				}
+				// Append intermediate diff to common diff.
+				for key, balanceDiff := range txDiff {
+					if err := commonDiff.appendBalanceDiffStr(key, balanceDiff); err != nil {
+						return nil, false, err
+					}
+				}
+			default:
+				return nil, false, errors.Errorf("unsupported script action '%T'", a)
 			}
-		default:
-			return txBalanceChanges{}, errors.Errorf("unsupported script action '%T'", a)
 		}
 	}
 	// Remove diffs from invoke stor.
 	ia.invokeDiffStor.invokeDiffsStor.reset()
 	// Add these diffs as a common diff to main stor.
-	if err := ia.saveDiff(commonDiff, info); err != nil {
-		return txBalanceChanges{}, err
+	if !ok {
+		err = ia.saveDiff(accountsChanges.diff, info)
+	} else {
+		err = ia.saveDiff(commonDiff, info)
+	}
+	if err != nil {
+		return nil, false, err
 	}
 	// Check transaction fee.
 	sponsorshipActivated, err := ia.stor.features.isActivated(int16(settings.FeeSponsorship))
 	if err != nil {
-		return txBalanceChanges{}, err
+		return nil, false, err
 	}
 	if !sponsorshipActivated {
 		// Minimum fee is not checked before sponsorship activation.
-		return totalChanges, nil
+		return totalChanges.addresses(), true, nil
 	}
 	minWavesFee := scriptExtraFee*scriptRuns + feeConstants[proto.InvokeScriptTransaction]*FeeUnit
 	wavesFee := tx.Fee
 	if tx.FeeAsset.Present {
 		wavesFee, err = ia.stor.sponsoredAssets.sponsoredAssetToWaves(tx.FeeAsset.ID, tx.Fee)
 		if err != nil {
-			return txBalanceChanges{}, errors.Wrap(err, "failed to convert fee asset to waves")
+			return nil, false, errors.Wrap(err, "failed to convert fee asset to waves")
 		}
 	}
 	if wavesFee < minWavesFee {
-		return txBalanceChanges{}, errors.Errorf("tx fee %d is less than minimum value of %d\n", wavesFee, minWavesFee)
+		return nil, false, errors.Errorf("tx fee %d is less than minimum value of %d\n", wavesFee, minWavesFee)
 	}
-	return totalChanges, nil
+	return totalChanges.addresses(), true, nil
```

Comment:
Looks like this function returns `true` when Invoke actually failed. Am I wrong about it?